File: src/agents/Support/connectivity_mapper.py
import logging
from pathlib import Path

logger = logging.getLogger(__name__)

class ConnectivityMapper:
    """Assistente Técnico: Especialista em Topologia e Acoplamento 🌐"""
    
    def calculate_metrics(self, file_path: str, data: dict, all_map: dict) -> dict:
        """Calcula instabilidade e acoplamento aferente/eferente."""
        eferent = len(data.get("dependencies", []))
        afferent = 0
        file_stem = Path(file_path).stem
        
        for other_file, other_data in all_map.items():
            if other_file == file_path: continue
            if any(file_stem in str(imp) for imp in other_data.get("dependencies", [])):
                afferent += 1
        
        instability = eferent / (afferent + eferent) if (afferent + eferent) > 0 else 0
        
        return {
            "in": afferent,
            "out": eferent,
            "instability": instability
        }

File: src_local/agents/Support/Analysis/connectivity_mapper.ts
import { HubManagerGRPC } from "../../../core/hub_manager_grpc";
import { Path } from "../../../core/path_utils.ts";
import winston from "winston";

const logger = winston.child({ module: "ConnectivityMapper" });

/**
 * 🌐 Mapeador de Conectividade PhD (gRPC Proxy).
 */
export class ConnectivityMapper {
    constructor(private hubManager?: HubManagerGRPC) { }

    /**
     * Calcula acoplamento de TODOS os arquivos em uma única passada via Hub Proxy.
     */
    async calculateBulk(allMap: Record<string, any>): Promise<Record<string, { in: number, out: number, instability: number }>> {
        if (!this.hubManager) {
            logger.warn("HubManager missing, bulk connectivity analysis skipped");
            return {};
        }

        try {
            const results = await this.hubManager.getConnectivity(allMap);
            const mapping: Record<string, any> = {};

            if (results && Array.isArray(results)) {
                results.forEach((r: any) => {
                    mapping[r.file] = { in: r.afferent, out: r.eferent, instability: r.instability };
                });
            }

            return mapping;
        } catch (err) {
            logger.error("gRPC connectivity analysis failed:", err);
            return {};
        }
    }

    calculateMetrics(filePath: string, data: any, allMap: Record<string, any>): any {
        const startTime = Date.now();
        const eferent = (data.dependencies || []).length;
        let afferent = 0;
        const fileStem = new Path(filePath).stem();

        for (const [otherFile, otherData] of Object.entries(allMap)) {
            if (otherFile === filePath) continue;
            const deps = otherData.dependencies || [];
            if (deps.some((imp: string) => imp.includes(fileStem))) {
                afferent += 1;
            }
        }

        const instability = (afferent + eferent) > 0 ? eferent / (afferent + eferent) : 0;
        const duration = (Date.now() - startTime) / 1000;
        logger.debug(`⏱️ [Connectivity] Single analysis for ${filePath} in ${duration.toFixed(4)}s`);

        return { in: afferent, out: eferent, instability: instability };
    }
}
